### ISSUE TITLE
2FA not supported

### ISSUE THREAD

**Issue by parkerlreed** [shadowmoose/RedditDownloader#22]
For the initial login I used \<password\>:\<auth-code\> to login. That was successful but it seems every time the script is invoked, Reddit is expecting a new auth code on the password. I'd rather not disable my 2FA for this.

````
[parker@inspiron15 RedditDownloader]$ python main.py

====================================
    Reddit Media Downloader 2.0
====================================
    (By ShadowMoose @ Github)

Loading all settings from file.
Loaded Source:  <source>
Authenticating via OAuth...
Traceback (most recent call last):
  File "main.py", line 227, in <module>
    p = Scraper(settings, custom_settings)
  File "main.py", line 116, in __init__
    reddit.login()
  File "./classes/reddit/reddit.py", line 36, in login
    _user = _reddit.user.me()
  File "/usr/lib/python3.6/site-packages/praw/models/user.py", line 95, in me
    user_data = self._reddit.get(API_PATH['me'])
  File "/usr/lib/python3.6/site-packages/praw/reddit.py", line 371, in get
    data = self.request('GET', path, params=params)
  File "/usr/lib/python3.6/site-packages/praw/reddit.py", line 486, in request
    params=params)
  File "/usr/lib/python3.6/site-packages/prawcore/sessions.py", line 182, in request
    params=params, url=url)
  File "/usr/lib/python3.6/site-packages/prawcore/sessions.py", line 113, in _request_with_retries
    data, files, json, method, params, retries, url)
  File "/usr/lib/python3.6/site-packages/prawcore/sessions.py", line 98, in _make_request
    params=params)
  File "/usr/lib/python3.6/site-packages/prawcore/rate_limit.py", line 32, in call
    kwargs['headers'] = set_header_callback()
  File "/usr/lib/python3.6/site-packages/prawcore/sessions.py", line 142, in _set_header_callback
    self._authorizer.refresh()
  File "/usr/lib/python3.6/site-packages/prawcore/auth.py", line 328, in refresh
    password=self._password)
  File "/usr/lib/python3.6/site-packages/prawcore/auth.py", line 142, in _request_token
    payload.get('error_description'))
prawcore.exceptions.OAuthException: invalid_grant error processing request
````

**Comment by shadowmoose** [comment]
Thanks for catching this one - I certainly don't want anybody disabling 2FA!

I'll take a look at what Reddit expects for 2FA auth, and roll out a better login setup.

**Comment by shadowmoose** [comment]
Okay, as an update, my optimal method of swapping to full oAuth is currently stuck awaiting a bugfix in PRAW. I've reached out to the developer to get things underway on that.

In the meantime, I may roll up a branch that can operate with a refresh token, but would require users to register a new "web app" with Reddit. That would at least let people use it for a while, since I have no idea how long the turnaround on this PRAW issue is going to be.

**Comment by aliparlakci** [comment]
You don't need people to create a web app. You, as personal, should create one app and enter the app's client_id and client_secret. Having each user create a web app is redundant.

**Comment by shadowmoose** [comment]
@aliparlakci That is correct, but not optimal for RMD. It requires the callback port to be limited to specifically the one I choose. RMD may not always be run on setups with that port exposed once the webui and containers are rolled out, and requiring users to expose the port for first-time setup isn't optimal.

The best way would be to authorize using a set of baked-in Installed Application credentials, but Praw seems to incorrectly implement that flow currently. Praw also doesn't function properly with tokens generated manually, so I've reached out the developer about the issue - but it's a slow process. 

If they haven't resolved the issue on their side by the time I'm pushing for the next release, it's a simple process to roll out the standard web app oath. I'd prefer to handle it all in one final auth update, through, rather than eventually rolling out another change down the road.

**Comment by shadowmoose** [comment]
I brought this issue to their attention, and Praw has finally fixed their support for oAuth via Installed Applications - as of version [6.0.0](https://praw.readthedocs.io/en/latest/package_info/change_log.html#id1). The next RMD release will have full oAuth support baked in.

**Comment by shadowmoose** [comment]
After a (really, really) long delay, this is finally officially resolved as of [Release 3.0.0](https://github.com/shadowmoose/RedditDownloader/releases/tag/3.0.0). 

Feel free to reopen if this still needs more attention.
